saleor/discount/utils.py

```diff
@@ -58,21 +61,28 @@ def get_product_discount_on_sale(
         or product_collections.intersection(discount.collection_ids)
     )
     if is_product_on_sale:
-        return discount.sale.get_discount()
+        sale_channel_listing = discount.sale.channel_listing.filter(  # type: ignore
+            channel=channel
+        ).first()
```

Comment:
Maby we we should add `sale_channel_listing` to `DiscountInfo`?
In the current implementation, the query will be executed for each product separately.